### PR TITLE
added nearly completed lab3

### PR DESCRIPTION
**Pull request by Yzoop**

### CODE REVIEW

File: Lab3/include/avarage_thr.h [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include "labArgument.h"
+
+void avarage(void* avaragearg);
```
Comment (pashatyl) [inline Lab3/include/avarage_thr.h]:
```suggestion
void average(void* avaragearg);
```

File: Lab3/src/avarage_thr.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    float sum = 0;
+    for (int i = 0; i < arg->arr_size; i++) {
+        sum += arg->arr[i];
+        count++;
+        Sleep(12);
+    }
+    arg->avarage = sum / count;
+    std::cout << "Thread average:\n";
+    std::cout << "\t" << arg->avarage << "\n";
+    _endthread();
```
Comment (pashatyl) [inline Lab3/src/avarage_thr.cpp]: Do not use _endthread
https://stackoverflow.com/questions/26960199/using-endthread

File: Lab3/include/avarage_thr.h (hunk 1 above)
Comment (pashatyl) [inline Lab3/include/avarage_thr.h]: add const qualifiers everywhere it's possible

Comment (pashatyl) [inline Lab3/include/avarage_thr.h]: please, use snake_case

File: Lab3/include/labArgument.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+
```
Comment (pashatyl) [inline Lab3/include/labArgument.h]: unnecessary empty line

File: Lab3/include/minmax_thr.h [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#pragma once
```
Comment (pashatyl) [inline Lab3/include/minmax_thr.h]: _thr in file name looks really strange, for such cases use folder named Threads instead

File: Lab3/src/avarage_thr.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    float count = 0;
+    float sum = 0;
+    for (int i = 0; i < arg->arr_size; i++) {
+        sum += arg->arr[i];
+        count++;
+        Sleep(12);
+    }
+    arg->avarage = sum / count;
+    std::cout << "Thread average:\n";
+    std::cout << "\t" << arg->avarage << "\n";
```
Comment (pashatyl) [inline Lab3/src/avarage_thr.cpp]: Unnecessary multiple calls

File: Lab3/src/avarage_thr.cpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#include "avarage_thr.h"
+#include <process.h>
+#include <Windows.h>
+#include <iostream>
+
+void avarage(void* avaragearg)
```
Comment (pashatyl) [inline Lab3/src/avarage_thr.cpp]: Too much responsibilities for function

File: Lab3/src/main.cpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
+﻿//#define _CRT_SECURE_NO_WARNINGS
```
Comment (pashatyl) [inline Lab3/src/main.cpp]: remove

File: Lab3/src/main.cpp [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+#include <process.h>
+#include <Windows.h>
+#include <iostream>
+#include "labArgument.h"
+#include "minmax_thr.h"
```
Comment (pashatyl) [inline Lab3/src/main.cpp:5]: different naming conventions in file naming

File: Lab3/src/main.cpp [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+#include "avarage_thr.h"
+
+
+using namespace std;
```
Comment (pashatyl) [inline Lab3/src/main.cpp]: use std:: explicitly

File: Lab3/src/main.cpp [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    avaragearg_t* avaragearg = get_new_avarage_structure(minmaxarg->arr, array_size, 0);
+    begin_avarge_thread(avaragearg);
+
+    minmaxarg->arr[minmaxarg->minId] = avaragearg->avarage;
+    minmaxarg->arr[minmaxarg->maxId] = avaragearg->avarage;
+
+
+    cout << "Result: \n";
+    for (int i = 0; i < array_size; i++) {
```
Comment (pashatyl) [inline Lab3/src/main.cpp:34]: using std::copy + std::backward_inserter will be prettier

File: Lab3/src/main.cpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+bool array_size_correct(int size);
+
+
+int main()
+{
+    int array_size = 0;
+    int* arr;
+
+    read_size_to(array_size);
+    arr = new int[array_size];
```
Comment (pashatyl) [inline Lab3/src/main.cpp:20]: use std::vector instead

File: Lab3/src/main.cpp [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    }
+    cout << endl;
+    
+    delete minmaxarg;
+    delete avaragearg;
+    delete[] arr;
+}
+
+
+void read_size_to(int& size)
```
Comment (pashatyl) [inline Lab3/src/main.cpp:45]: It would looks better in separated module

File: Lab3/src/main.cpp [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+{
+    cout << "Please enter array elements: \n";
+    for (int i = 0; i < size; i++) {
+        cout << "Reading " << i + 1 << "/" << size << " element: ";
+        cin >> to_arr[i];
+    }
+}
+
+
+void begin_minmax_thread(const minmaxarg_t* minmaxarg)
```
Comment (pashatyl) [inline Lab3/src/main.cpp:65]: also, to separated module

File: Lab3/src/main.cpp [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+{
+    cout << "Starting minmax thread.\nPlease, wait...";
+    HANDLE h_minmax = (HANDLE)_beginthread(min_max, 0, (void*)minmaxarg);
+    WaitForSingleObject(h_minmax, INFINITE);
+    cout << "\n------------------------\nMinmax thread has finished.\n";
+
+}
+
+
+void begin_avarge_thread(const avaragearg_t* avaragearg)
```
Comment (pashatyl) [inline Lab3/src/main.cpp]: Code duplication

File: Lab3/src/minmax_thr.cpp [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        if (max < arg->arr[i]) {
+            max = arg->arr[i];
+            arg->maxId = i;
+        }
+        if (min > arg->arr[i]) {
+            min = arg->arr[i];
+            arg->minId = i;
+        }
+        Sleep(7);
```
Comment (pashatyl) [inline Lab3/src/minmax_thr.cpp]: magic number
Introduce define for enabling this testing call

File: Lab3/src/minmax_thr.cpp [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+#include <process.h>
+
+#include "minmax_thr.h"
+
+using namespace std;
+
+void min_max(void* minmaxarg)
+{
+    minmaxarg_t* arg = (minmaxarg_t*)minmaxarg;
+    int max = -INT_MAX;
```
Comment (pashatyl) [inline Lab3/src/minmax_thr.cpp]: why not INT_MIN? are you sure INT_MAX == - INT_MIN?

File: Lab3/include/labArgument.h [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+struct minmaxarg_t 
+{
+	int* arr = nullptr;
+	int arr_size, min, max, minId, maxId;
```
Comment (pashatyl) [inline Lab3/include/labArgument.h]: use multiple lines

File: Lab3/src/minmax_thr.cpp [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    }
+    int min_global = min;
+    int max_global = max;
+    cout << "Thread min_max: \n";
+    cout << "\tMax: " << max << "\n\tMin: " << min << "\n";
+    _endthread();
+}
+
+
+minmaxarg_t* get_new_minmax_structure(int* arr, int arr_size, int min, int max, int minid, int maxid)
```
Comment (pashatyl) [inline Lab3/src/minmax_thr.cpp]: isn't it a constructor?)

**Review by pashatyl — CHANGES_REQUESTED** [review]
Some code style was checked, behavior will be reviewed after fixes

File: Lab3/src/minmax_thr.cpp (hunk 18 above)
Comment (kalinchukivan) [reply Lab3/src/minmax_thr.cpp]: First steps of OOP

File: Lab3/include/avarage_thr.h (hunk 1 above)
Comment (Yzoop) [reply Lab3/include/avarage_thr.h]: An argument avaragearg is supposed to be changed while avarage() func is working

**Pull request closed without merging**